src/core/libmaven/EIC.cpp

```diff
@@ -124,29 +128,120 @@ EIC *EIC::eicMerge(const vector<EIC *> &eics)
+        return false;
+    }
+
+    return true;
+}
+
+void EIC::_computeAsLSBaseline(float lambda, float p, int numIterations)
+{
+    if (!_clearBaseline())
+        return;
+
+    using namespace Eigen;
+    unsigned int n = static_cast<unsigned int>(intensity.size());
+
+    // create a sparse identity matrix of size `n`
+    auto ident = MatrixXf::Identity(n, n).sparseView();
+
+    // compute approximate second derivative of the identity matrix
+    auto D = _diff(_diff(ident));
+
+    // create a weights vector of length `n` initially containing only ones
+    // for coefficients
+    auto w = VectorXf(ArrayXf::Ones(n));
+
+    // instantiate a Cholesky decomposition linear solver for sparse matrices
+    SimplicialCholesky<SparseMatrix<float>> solver;
+
+    // instantiate a vector for intensity and one to store baseline
+    VectorXf intensityVec = Map<VectorXf>(intensity.data(), n);
+    VectorXf baselineVec;
+
+    // helper matrices for `select` operation used when creating binary vectors
+    auto ones = VectorXf(MatrixXf::Ones(n, 1));
+    auto zeros = VectorXf(MatrixXf::Zero(n, 1));
+
+    // TODO: ideally this should converge to a point where the baseline does
+    // not change anymnore, but since we do not have good float comparators
+    // yet we can use a decent number of iterations to get as close to the true
+    // baseline as possible
+    for (int i=0; i < numIterations; ++i) {
+        // create a square matrix 'W' with the weights as its diagonal
+        auto W = SparseMatrix<float>(w.asDiagonal());
+
+        // compute 'A' and 'b', and then solve for 'x' that satisfies the
+        // equation 'A·x = b', where x will be the iteratively estimated
+        // baseline
+        auto A = W + (lambda * (D.transpose() * D));
+        solver.compute(A);
+        auto b = VectorXf(w.array() * intensityVec.array());
+        baselineVec = solver.solve(b);
+
+        // calculate weights for the next iteration
+        auto gtBin = VectorXf(
+            ((intensityVec - baselineVec).array() > 0.0f).select(ones, zeros)
+        );
+        auto ltBin = VectorXf(
+            ((intensityVec - baselineVec).array() < 0.0f).select(ones, zeros)
+        );
+        w = (p * gtBin) + ((1.0f - p) * ltBin);
     }
 
+    // copy data from an Eigen vector to std::vector
+    vector<float> clippedBaseline(baselineVec.data(),
+                                  baselineVec.data()
+                                  + baselineVec.rows()
+                                  * baselineVec.cols());
+
+    // clip negative values from the baseline vector
+    std::transform(begin(clippedBaseline),
+                   end(clippedBaseline),
+                   begin(clippedBaseline),
+                   [](float f) { return f < 0.0f ? 0.0f : f; });
+
+    // since baseline right now is an array of float, the data from clipped
+    // vector is copied to this array
+    std::copy(begin(clippedBaseline), end(clippedBaseline), baseline);
+}
```

Comment:
I do not think the reader (for this piece of code) should have to go through each step and affirm that the return type on each step is correct, or even be aware of it. Most of these variables are some type derived from templates of Eigen library's `Matrix` or `Vector` meta types. Putting them within this already complicated algorithm will in fact decrease the readability in my opinion. Also I believe `auto` was meant for being used as a [duck typing](https://en.wikipedia.org/wiki/Duck_typing) alternative to hide precisely this kind of type complexity.

> I also feel some of the variable declared above can be named better.

Apart from `w` (for weights), I think the rest of the variables are fine. I myself am usually very adamant about verbosity of variables, but since this algorithm was derived from the original author's pseudo-code implementation, the variables have been kept as such to allow easy identification of all elements involved. Besides its a common practice (in mathematical computing) to use single letter lower-case symbols for vectors and upper-case symbols for Matrices. Additionally, the formula `A·x = b` is extremely commonly used for representing a system of linear equations, and hence the name of the symbols.

As for the comments, I think, a few months down the line developers (including me) will have an easier time debugging this where each step for this algorithm is slightly explained. Please allow this exception for such functions implementing scientific papers.